### PR TITLE
[Merged by Bors] - feat: SameCycle.equivalence and SameCycle.setoid

### PR DESCRIPTION
**Pull request by lambda-fairy**
I'm exploring @kmill's combinatorial maps work and these two definitions are part of it:

https://github.com/kmill/lean-graphs/blob/bd3139daf81c0f41ce5ade951195858cdf038bd2/src/myequiv.lean#L72-L76

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean [hunk 1]
```diff
@@ -100,6 +100,12 @@ theorem SameCycle.trans : SameCycle f x y → SameCycle f y z → SameCycle f x
   fun ⟨i, hi⟩ ⟨j, hj⟩ => ⟨j + i, by rw [zpow_add, mul_apply, hi, hj]⟩
 #align equiv.perm.same_cycle.trans Equiv.Perm.SameCycle.trans
 
+theorem SameCycle.equivalence : Equivalence (SameCycle f) :=
+  ⟨SameCycle.refl f, SameCycle.symm, SameCycle.trans⟩
+
+def SameCycle.setoid (f : Perm α) : Setoid α :=
+  ⟨SameCycle f, SameCycle.equivalence⟩
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/Perm/Cycle/Basic.lean:107]:
```suggestion
def SameCycle.setoid (f : Perm α) : Setoid α where
  iseqv := SameCycle.equivalence
```

**Review by eric-wieser — APPROVED** [review]

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean (hunk 1 above)
Comment (lambda-fairy) [reply Mathlib/GroupTheory/Perm/Cycle/Basic.lean:107]: Thanks for the suggestion – I tried that locally and got:

```
don't know how to synthesize implicit argument
  @equivalence α (?m.33592 setoid f)
context:
ι: Type u_1
α: Type u_2
β: Type u_3
f✝g: Perm α
p: α → Prop
xyz: α
f: Perm α
⊢ Perm α
```

I guess Lean can't figure out the `f` 🤔

Comment (lambda-fairy) [reply Mathlib/GroupTheory/Perm/Cycle/Basic.lean:107]: `@SameCycle.equivalence _ f` does work. Though if we're going down that route, should we just make the `f` on `SameCycle.equivalence` an explicit parameter instead?

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean [hunk 2]
```diff
@@ -100,6 +100,12 @@ theorem SameCycle.trans : SameCycle f x y → SameCycle f y z → SameCycle f x
   fun ⟨i, hi⟩ ⟨j, hj⟩ => ⟨j + i, by rw [zpow_add, mul_apply, hi, hj]⟩
 #align equiv.perm.same_cycle.trans Equiv.Perm.SameCycle.trans
 
+theorem SameCycle.equivalence : Equivalence (SameCycle f) :=
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/Perm/Cycle/Basic.lean:103]:
```suggestion
variable (f) in
theorem SameCycle.equivalence : Equivalence (SameCycle f) :=
```

**Review by eric-wieser — APPROVED** [review]
bors d+

Thanks!

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean (hunk 2 above)
Comment (lambda-fairy) [reply Mathlib/GroupTheory/Perm/Cycle/Basic.lean:104]: TIL about `variable _ in`, thanks!

**Comment by lambda-fairy** [comment]
Added a doc string to `SameCycle.setoid`. Apologies!

bors r+

**Pull request closed without merging**